front/temporal/agent_schedule/activities.ts

```diff
@@ -1,34 +1,78 @@
+import {
+  isMCPConfigurationForRunAgent,
+  isServerSideMCPServerConfiguration,
+} from "@app/lib/actions/types/guards";
 import { getAgentConfiguration } from "@app/lib/api/assistant/configuration/agent";
 import {
   createConversation,
   postUserMessage,
 } from "@app/lib/api/assistant/conversation";
 import type { AuthenticatorType } from "@app/lib/auth";
 import { Authenticator } from "@app/lib/auth";
+import { ConversationResource } from "@app/lib/resources/conversation_resource";
+import { MCPServerViewResource } from "@app/lib/resources/mcp_server_view_resource";
+import { TriggerResource } from "@app/lib/resources/trigger_resource";
 import logger from "@app/logger/logger";
+import type { AgentConfigurationType } from "@app/types";
 import type { TriggerType } from "@app/types/assistant/triggers";
 
-export async function runScheduledAgentsActivity(
-  authType: AuthenticatorType,
-  trigger: TriggerType
-) {
-  const auth = await Authenticator.fromJSON(authType);
-
-  if (!auth.workspace() || !auth.user()) {
-    throw new Error("Invalid authentication. Missing workspaceId or userId.");
+/**
+ * We want to create individual conversations if the agent outcome will vary from user to user.
+ */
+async function shouldCreateIndividualConversations(
+  auth: Authenticator,
+  agentConfiguration: AgentConfigurationType,
+  checkedAgentConfigurationIds: string[] = []
+): Promise<boolean> {
+  // Check if one of the actions is using a personal actions or a run agent action
+  const mcpServerViews = await MCPServerViewResource.listByWorkspace(auth);
+  const mcpServerViewsMap = new Map(
+    mcpServerViews.map((mcpServerView) => [mcpServerView.sId, mcpServerView])
+  );
+  for (const action of agentConfiguration.actions) {
+    if (isServerSideMCPServerConfiguration(action)) {
+      const mcpServerView = mcpServerViewsMap.get(action.mcpServerViewId);
+      if (!mcpServerView) {
+        throw new Error(
+          `MCP server view with ID ${action.mcpServerViewId} not found.`
+        );
+      }
+      if (mcpServerView.oAuthUseCase === "personal_actions") {
+        return true;
+      }
+      // Check the chain of agents
+      if (
+        isMCPConfigurationForRunAgent(action) &&
+        action.childAgentId &&
+        // Avoid infinite loop
+        !checkedAgentConfigurationIds.includes(action.childAgentId)
+      ) {
+        const subAgentConfiguration = await getAgentConfiguration(auth, {
+          agentId: action.childAgentId,
+          variant: "full",
+        });
+        if (subAgentConfiguration) {
+          const subCheck = await shouldCreateIndividualConversations(
+            auth,
+            subAgentConfiguration,
+            [...checkedAgentConfigurationIds, agentConfiguration.sId]
+          );
+          if (subCheck) {
+            return true;
+          }
```

Comment:
```suggestion
          return shouldCreateIndividualConversations(
            auth,
            subAgentConfiguration,
            [...checkedAgentConfigurationIds, agentConfiguration.sId]
          );
```

can we just return shouldCreateIndividualConversations like this ?

Reply:
ah! true :) missed that